pkg/server/server.go

```diff
@@ -476,3 +646,264 @@ func adaptContext(parent genericapiserver.PostStartHookContext) context.Context
+		}
+
+		return nil, err
+	}
+
+	if workspace.Spec.InheritFrom == "" {
+		return nil, apierrors.NewNotFound(schema.GroupResource{Group: apiextensionsv1.SchemeGroupVersion.Group, Resource: "customresourcedefinitions"}, "")
+	}
+
+	sourceWorkspaceKey := clusters.ToClusterAwareKey(adminClusterName, workspace.Spec.InheritFrom)
+	if _, err := c.workspaceLister.Get(sourceWorkspaceKey); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, apierrors.NewNotFound(schema.GroupResource{Group: apiextensionsv1.SchemeGroupVersion.Group, Resource: "customresourcedefinitions"}, "")
+		}
+
+		return nil, err
+	}
+
+	sourceWorkspaceCRDKey := clusters.ToClusterAwareKey(workspace.Spec.InheritFrom, name)
+	crd, err = c.crdLister.Get(sourceWorkspaceCRDKey)
+	return crd, err
+}
+
+func findCRD(crdName string, crds []*apiextensionsv1.CustomResourceDefinition) (*apiextensionsv1.CustomResourceDefinition, bool) {
+	var crd *apiextensionsv1.CustomResourceDefinition
+
+	for _, aCRD := range crds {
+		if aCRD.Name != crdName {
+			continue
+		}
+		if crd == nil {
+			crd = aCRD
+		} else {
+			if !equality.Semantic.DeepEqual(crd.Spec, aCRD.Spec) {
+				return crd, false
+			}
+		}
+	}
+
+	return crd, true
+}
+
+type kcpAPIExtensionsSharedInformerFactory struct {
+	apiextensionsinformers.SharedInformerFactory
+	workspaceLister tenancylisters.WorkspaceLister
+}
+
+func (f *kcpAPIExtensionsSharedInformerFactory) Apiextensions() apiextensions.Interface {
+	i := f.SharedInformerFactory.Apiextensions()
+	return &kcpAPIExtensionsApiextensions{
+		Interface:       i,
+		workspaceLister: f.workspaceLister,
+	}
+}
+
+type kcpAPIExtensionsApiextensions struct {
+	apiextensions.Interface
+	workspaceLister tenancylisters.WorkspaceLister
+}
+
+func (i *kcpAPIExtensionsApiextensions) V1() apiextensionsinformerv1.Interface {
+	v1i := i.Interface.V1()
+	return &kcpAPIExtensionsApiextensionsV1{
+		Interface:       v1i,
+		workspaceLister: i.workspaceLister,
+	}
+}
+
+type kcpAPIExtensionsApiextensionsV1 struct {
+	apiextensionsinformerv1.Interface
+	workspaceLister tenancylisters.WorkspaceLister
+}
+
+func (i *kcpAPIExtensionsApiextensionsV1) CustomResourceDefinitions() apiextensionsinformerv1.CustomResourceDefinitionInformer {
+	c := i.Interface.CustomResourceDefinitions()
+	return &kcpAPIExtensionsApiextensionsV1CustomResourceDefinitionInformer{
+		CustomResourceDefinitionInformer: c,
+		workspaceLister:                  i.workspaceLister,
+	}
+}
+
+type kcpAPIExtensionsApiextensionsV1CustomResourceDefinitionInformer struct {
+	apiextensionsinformerv1.CustomResourceDefinitionInformer
+	workspaceLister tenancylisters.WorkspaceLister
+}
+
+func (i *kcpAPIExtensionsApiextensionsV1CustomResourceDefinitionInformer) Lister() apiextensionslisters.CustomResourceDefinitionLister {
+	originalLister := i.CustomResourceDefinitionInformer.Lister()
+	l := &kcpCRDLister{
+		crdLister:       originalLister,
+		workspaceLister: i.workspaceLister,
+	}
+	return l
+}
+
+// COPIED FROM kube-aggregator
```

Comment:
@smarterclayton what's the best way to do attribution here or what should I do?

Reply:
simple enough code, no? Not complex enough to fall under attribution requirements ;-)

Reply:
Agreed, just checking